### PR TITLE
Topic/dc mlx5

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1543/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2205/ for details.

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1570/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2237/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1574/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2241/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2242/ for details.

**Comment by alex-mikheev** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2243/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1575/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2244/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1576/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2245/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1577/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2265/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1593/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2266/ for details.

**Comment by alex-mikheev** [comment]
@yosefe everything looks green. ping!

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2309/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1634/ for details (Mellanox internal link).

File: src/uct/ib/dc/accel/dc_mlx5.c [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+
+static ucs_status_t uct_dc_mlx5_iface_query(uct_iface_h tl_iface, uct_iface_attr_t *iface_attr)
+{
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_iface, uct_dc_mlx5_iface_t);
+
+    uct_dc_iface_query(&iface->super, iface_attr);
+    uct_rc_mlx5_iface_common_query(&iface->super.super, iface_attr, IBV_EXP_QPT_DC_INI);
+
+    /*TODO: remove flags once we have a full functionality */
+    iface_attr->cap.flags           = UCT_IFACE_FLAG_AM_SHORT|
```
Comment (yosefe) [inline src/uct/ib/dc/accel/dc_mlx5.c]: code style - space before |

File: src/uct/ib/dc/accel/dc_mlx5.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+typedef struct uct_dc_mlx5_iface { 
+    uct_dc_iface_t              super;
+    uct_rc_mlx5_iface_common_t  mlx5_common;
+    uct_ib_mlx5_txwq_t         *dci_wqs;
+} uct_dc_mlx5_iface_t;
+
+typedef struct uct_dc_mlx5_ep {
+    uct_dc_ep_t          super;
+    struct mlx5_wqe_av   av;
+} uct_dc_mlx5_ep_t;
```
Comment (yosefe) [inline src/uct/ib/dc/accel/dc_mlx5.h]: pls add it to type_info.c to print he size of this struct

File: src/uct/ib/dc/accel/dc_mlx5.c [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+uct_dc_mlx5_poll_tx(uct_dc_mlx5_iface_t *iface) 
+{
+    uint8_t dci;
+    struct mlx5_cqe64 *cqe;
+    uint32_t qp_num;
+    uint16_t hw_ci;
+    UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
+
+    cqe = uct_ib_mlx5_get_cqe(&iface->super.super.super, &iface->mlx5_common.tx.cq,
+            UCT_IB_MLX5_CQE64_SIZE_LOG);
```
Comment (yosefe) [inline src/uct/ib/dc/accel/dc_mlx5.c]: indentation

File: src/uct/ib/dc/accel/dc_mlx5.c [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+
+    status = uct_rc_mlx5_iface_common_init(&self->mlx5_common, &self->super.super,
+                                           &config->super.super);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    self->dci_wqs = ucs_malloc(self->super.tx.ndci * sizeof(uct_ib_mlx5_txwq_t), "dc");
+    if (self->dci_wqs == NULL) {
+        return UCS_ERR_NO_MEMORY;
```
Comment (yosefe) [inline src/uct/ib/dc/accel/dc_mlx5.c]: need to cleanup what uct_rc_mlx5_iface_common_init() did - e.g mpool

File: src/uct/ib/dc/base/dc_iface.h [hunk 5]
```diff
@@ -101,4 +101,27 @@ static inline ucs_arbiter_t *uct_dc_iface_dci_waitq(uct_dc_iface_t *iface)
...
+    uct_rc_txqp_t *txqp;
+
+    txqp = &iface->tx.dcis[dci].txqp;
+
+    if (uct_rc_txqp_available(txqp) == (int16_t)iface->super.config.tx_qp_len) {
+        return UCS_OK;
+    }
+    ucs_trace_data("dci %d is not flushed %d/%d", dci, 
+                    txqp->available, iface->super.config.tx_qp_len);
+    if (uct_rc_txqp_unsignaled(txqp) != 0) {
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.h]: this should be an assertion because in DC we always send signaled

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 6]
```diff
@@ -312,6 +347,37 @@ uct_ib_mlx5_get_next_seg(uct_ib_mlx5_txwq_t *wq, void *seg_base, size_t seg_len)
     return rseg;
 }
 
+static unsigned UCS_F_ALWAYS_INLINE uct_ib_mlx5_ctrl_seg_size(int qp_type)
+{
+    switch (qp_type) {
+        case IBV_QPT_RC:
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.h]: code style - case and switch on same column

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        srq->ready_idx        = index;
+        srq->sw_pi            = index;
+        iface->rx.available  -= count;
+        ucs_memory_cpu_store_fence();
+        *srq->db = htonl(srq->sw_pi);
+        ucs_assert(uct_ib_mlx5_srq_get_wqe(srq, srq->mask)->srq.next_wqe_index == 0);
+    }
+    return count;
+}
+
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: code style: single space between functions in C source file

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        index = next;
+    }
+    srq->tail = index;
+}
+
+ucs_status_t uct_rc_mlx5_iface_common_init(uct_rc_mlx5_iface_common_t *iface, uct_rc_iface_t *rc_iface,
+                                           uct_rc_iface_config_t *config)
+{
+    ucs_status_t status;
+
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: single space

File: src/uct/ib/rc/accel/rc_mlx5_common.h [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+done:
+    max_batch = rc_iface->super.config.rx_max_batch;
+    if (rc_iface->rx.available >= max_batch) {
+        uct_rc_mlx5_iface_srq_post_recv(rc_iface, &mlx5_common_iface->rx.srq);
+    }
+    return status;
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_rc_mlx5_txqp_post_send(uct_rc_iface_t *iface, uct_rc_txqp_t *txqp, uct_ib_mlx5_txwq_t *txwq,
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.h]: pls rename to uct_rc_mlx5_common_post_send

File: test/gtest/uct/uct_test.h [hunk 10]
```diff
@@ -179,6 +179,7 @@ std::ostream& operator<<(std::ostream& os, const resource* resource);
     rc_mlx5,            \
     rc,                 \
     dc,                 \
+    dc_mlx5,            \
```
Comment (yosefe) [inline test/gtest/uct/uct_test.h]: need to add to ucp tests as well

File: src/uct/ib/dc/accel/dc_mlx5.c [hunk 11]
```diff
@@ -0,0 +1,655 @@
...
+{
+    uct_dc_mlx5_iface_t *iface = arg;
+    ucs_status_t status;
+
+    status = uct_rc_mlx5_iface_common_poll_rx(&iface->mlx5_common, &iface->super.super);
+    if (status == UCS_ERR_NO_PROGRESS) {
+        uct_dc_mlx5_poll_tx(iface);
+    }
+}
+
```
Comment (yosefe) [inline src/uct/ib/dc/accel/dc_mlx5.c]: single space line

**Comment by yosefe** [comment]
@alex-mikheev looks good except minor comments, also we should add these optimizations to TODO:
sc2cqe
* Use shrink AV
* O(1) dci lookup
* use flow control for DC
* allocate uct_dc_mlx5_iface::dci_wqs inline instead of pointer to (reduce cache misses)
* optimize memory footprint of uct_dc_mlx5_ep_t
* register dc progress on demand

**Comment by alex-mikheev** [comment]
DC TODO: https://github.com/openucx/ucx/issues/942

File: src/uct/ib/dc/accel/dc_mlx5.c (hunk 4 above)
Comment (alex-mikheev) [inline src/uct/ib/dc/accel/dc_mlx5.c]: this is done in uct_rc_mlx5_iface_common_cleanup()

Comment (yosefe) [inline src/uct/ib/dc/accel/dc_mlx5.c]: but i don't see it's called on error flow...

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2324/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1649/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1: please squash

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1655/ for details (Mellanox internal link).

**Pull request merged**